rts/motoko-rts/src/idl.rs

```diff
@@ -165,32 +197,34 @@ unsafe fn parse_idl_header<M: Memory>(
             // internal
             // See Note [mutable stable values] in codegen/compile.ml
             let t = sleb128_decode(buf);
-            check_typearg(extended, t, n_types);
+            check_typearg(mode, t, n_types);
         } else if ty >= 0 {
             idl_trap_with("illegal type table"); // illegal
-        } else if is_primitive_type(extended, ty) {
+        } else if is_primitive_type(mode, ty) {
             // illegal
             idl_trap_with("primitive type in type table");
         } else if ty == IDL_CON_opt {
             let t = sleb128_decode(buf);
-            check_typearg(extended, t, n_types);
+            check_typearg(mode, t, n_types);
         } else if ty == IDL_CON_vec {
             let t = sleb128_decode(buf);
-            check_typearg(extended, t, n_types);
+            check_typearg(mode, t, n_types);
         } else if ty == IDL_CON_record {
-            parse_fields(extended, buf, n_types);
+            parse_fields(mode, buf, n_types);
+        } else if ty == IDL_EXT_tuple {
```

Comment:
I suspect `IDL_EXT_tuple` should be rejected unless `mode == MemoryCompatibility` (which it never is in this function, see above).